autode/opt/optimisers/base.py

```diff
@@ -466,14 +465,213 @@ def final_coordinates(self):
         raise RuntimeError("A NullOptimiser has no coordinates")
 
 
+@dataclass
+class ConvergenceParams:
+    """
+    Various convergence parameters for optimisers and some common
+    preset convergence tolerances
+
+    Args:
+        abs_d_e: Absolute change in energy, |E_i - E_i-1|
+        rms_g: RMS of the gradient, RMS(∇E)
+        max_g: Maximum component of gradient, max(∇E)
+        rms_s: RMS of the last step, RMS(x_i - x_i-1)
+        max_s: Maximum component of last step, max(x_i - x_i-1)
+        strict: Whether all criteria must be converged strictly.
+                If False, convergence is signalled when some criteria
+                are overachieved and others are close to convergence
+
+    """
+
+    abs_d_e: Optional[PotentialEnergy] = None
+    rms_g: Optional[GradientRMS] = None
+    max_g: Optional[GradientRMS] = None
+    rms_s: Optional[Distance] = None
+    max_s: Optional[Distance] = None
+    strict: bool = False
+
+    def __post_init__(self):
+        """Type checking and sanity checks on parameters"""
+        self._num_attrs = ["abs_d_e", "rms_g", "max_g", "rms_s", "max_s"]
```

Comment:
@t-young31 I made this a property, but still kept hard-coded strs because the attributes may be None at the time of evaluation. I am not really sure how to implement something based on the type without convoluted code.